geant4: 10.4.1 -> 10.5.1

OmnipotentEntity opened the pull request:
###### Motivation for this change

Update to latest version (several months old at this point).  Update the data files to their latest version as well (some have been renamed and might disappear soon.)

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @Tmplt

OmnipotentEntity committed:
geant4: 10.4.1 -> 10.5.1

OmnipotentEntity commented:
I can't figure out how or even if the environment variables set by the packages herein worked previously.  I may just be doing something wrong.  You can set them manually and it works fine, but I wanted @tmplt feedback before modifying things with that, just in case I was screwing up.

veprbl commented:
@OmnipotentEntity The datasets are designed to work from `nix-shell` and from within nix builds. There is nothing put in place to make them automatically work in the static user environments (like `nix-env` or NixOS `systemPackages`).

Tmplt commented on pkgs/development/libraries/physics/geant4/datasets.nix:
```diff
@@ -27,6 +37,7 @@ let
         homepage = "https://geant4.web.cern.ch/support/download";
         license = licenses.g4sl;
         platforms = platforms.all;
+        inherit priority;
```
This I haven't seen before. What effect does different values have?

OmnipotentEntity replied:
It makes it so that the files that are named the same between the packages do not conflict.  Global default value is 5.   0 is highest priority.

Tmplt replied:
Alright, lgtm then.

veprbl replied:
Setting `meta.priority` will not fix the underlying issue. Let's not do that.

OmnipotentEntity commented:
@veprbl I figured as much from the shellhook statement, but I couldn't figure out the incantation to make nix-shell install the data packages.  I tried `nix-shell -p 'geant4 geant4.data` but I receive a big fat error message for my troubles.

veprbl commented on pkgs/development/libraries/physics/geant4/datasets.nix:
```diff
@@ -57,20 +68,22 @@ in
       version = "5.2";
```
Official web site advertises G4RadioactiveDecay5.3

OmnipotentEntity replied:
whoops, I did download 5.3, but because of the kerfluffle with the conflict between it and photonevaporation, I forgot to up the version number.  I'll force push a fix.

veprbl commented on pkgs/development/libraries/physics/geant4/datasets.nix:
```diff
@@ -1,22 +1,32 @@
 { stdenv, fetchurl, }:
```
Need to add G4TENDL1.3.2

veprbl replied:
Strangely enough I could not find the corresponding environment variable we would have to set for it. Perhaps we can skip this data for now.

OmnipotentEntity replied:
I couldn't find it either, which is why I skipped it.  I grepped through the geant4 code as well, and I don't know if it's even used?

veprbl replied:
Let's assume it is not needed.

veprbl reviewed (commented):
This update breaks `g4py` package. I guess it needs to be updated too.

Also, I suggest we take the changes adding `meta.priority` and "fixConflicts" out of this PR to focus on bumping the version first.

veprbl commented:
> @veprbl I figured as much from the shellhook statement, but I couldn't figure out the incantation to make nix-shell install the data packages. I tried `nix-shell -p 'geant4 geant4.data` but I receive a big fat error message for my troubles.

The "-p" option of nix-shell will not take the `geant4.data` because it's an attribute set (this limitation has to do with how this option is implemented in nix-shell). However, it should be possible to include specific datasets:
```
nix-shell -p geant4 -p geant4.data.G4NDL -p geant4.data.G4ABLA
```

veprbl commented on pkgs/development/libraries/physics/geant4/datasets.nix:
```diff
@@ -1,22 +1,32 @@
+          fi
+        }
+        fixConflicts README
+        fixConflicts History
```
A more robust approach would be to instead change installation path from "$out/data" to something of the form like "$out/share/Geant4-10.5.1/data/G4ABLA3.1" (I took this path from the "bin/geant4.sh").

veprbl replied:
(Such change would probably deserve it's own PR)

OmnipotentEntity replied:
that's a great idea!

OmnipotentEntity commented:
Should I also update `g4py` in this pull request, or should I issue a second one for it separately?

OmnipotentEntity commented:
This push fixes the issues listed above.  Including the one that @veprbl thought deserved its own PR, because I'd rather fix the underlying issue, and it's not that complicated a change.

veprbl commented on pkgs/development/libraries/physics/geant4/datasets-hook.sh:
```diff
@@ -1,5 +1,5 @@
 @name@ () {
-    export G4@envvar@DATA="@out@/data"
+    export G4@envvar@DATA="@out@/share/@pname@-@pver@/data/@name@@version@"
```
```suggestion
    export G4@envvar@DATA="@datadir@"
```

veprbl commented on pkgs/development/libraries/physics/geant4/datasets.nix:
```diff
@@ -1,22 +1,25 @@
       dontBuild = true;
       dontConfigure = true;
 
+      # Fix conflict between READMEs of various packages
```
This comment is not needed.

veprbl commented on pkgs/development/libraries/physics/geant4/datasets.nix:
```diff
@@ -1,22 +1,25 @@
       installPhase = ''
-        mkdir -p $out/data
-        mv ./* $out/data
+        OUTDIR=$out/share/${pname}-${pver}/data/${name}${version}
```
This line can be moved out to the derivation:
```nix
stdenv.mkDerivation {
  # ...
  datadir = "${placeholder "out"}/share/Geant4-${geant_version}/data/${name}${version}";
  installPhase = ''
    mkdir "$datadir"
    mv ./* "$datadir"
  '';
  # ...
}
```
Then it can be used as `@datadir@` in the setup hook, it will be also available for introspection (e.g. `geant4.data.G4NDL.datadir`) from different derivations.
I propose we hardcode the "Geant4" bit. Also could rename `pver` to something more readable like `geant_version`

veprbl commented on pkgs/development/libraries/physics/geant4/default.nix:
```diff
@@ -35,13 +35,13 @@
-  pname = "geant4";
+stdenv.mkDerivation rec {
+  version = "10.5.1";
+  pname = "Geant4";
```
Let's not adjust this.

veprbl commented on pkgs/development/libraries/physics/geant4/default.nix:
```diff
@@ -78,7 +78,10 @@ stdenv.mkDerivation {
-    data = import ./datasets.nix { inherit stdenv fetchurl; };
+    data = import ./datasets.nix {
+          inherit stdenv fetchurl pname;
+          pver = version;
```
```suggestion
          geant_version = version;
```

veprbl commented on pkgs/development/libraries/physics/geant4/default.nix:
```diff
@@ -78,7 +78,10 @@ stdenv.mkDerivation {
   passthru = {
-    data = import ./datasets.nix { inherit stdenv fetchurl; };
+    data = import ./datasets.nix {
+          inherit stdenv fetchurl pname;
```
```suggestion
          inherit stdenv fetchurl;
```

veprbl commented:
> Should I also update `g4py` in this pull request, or should I issue a second one for it separately?

I don't mind either way.

OmnipotentEntity commented:
I corrected the things you've indicated and updated g4py.  Tested it as working at the barest level (an otherwise empty python file with "import Geant4").

OmnipotentEntity commented:
Thanks @veprbl for all of the help and the reviews.  I think I learned a lot from you.

Thanks @Tmplt for the review and putting together this package in the first place ;D

veprbl commented on pkgs/development/libraries/physics/geant4/datasets.nix:
```diff
@@ -1,22 +1,24 @@
 
 let
-  mkDataset = { name, version, sha256, envvar}:
+  mkDataset = { name, version, sha256, envvar }:
```
👍

OmnipotentEntity replied:
oh gosh, hahaha

veprbl commented:
@GrahamcOfBorg build geant4 g4py

veprbl reviewed (approved):
Tested on x86_64-linux (non-nixos). The B1 example compiles and seems to work.

veprbl commented:
Thank you for your contribution! If there there are some questions left, feel free to open an issue or post on Discourse.
